### PR TITLE
[core] Introduce LocationAttr (unknown only)

### PR DESCRIPTION
**Pull request by Moxinilian**
This commit introduces basic support for location attributes. This only
supports unknown locations for now.

Printers and parsers now support locations on arbitrary operations and
block arguments, along with locations as attributes.

### CODE REVIEW

**Comment by Moxinilian** [comment]
Codecov this was a draft PR...

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -292,6 +292,16 @@ class UnitAttr(ParametrizedAttribute):
     name = "unit"
 
 
+@irdl_attr_definition
+class LocationAttr(ParametrizedAttribute):
```
Comment (math-fehr) [inline xdsl/dialects/builtin.py:296]: Can you add it to the `Builtin` dialect at the bottom of the file?

File: xdsl/xdsl_opt_main.py [hunk 2]
```diff
@@ -329,7 +329,11 @@ def register_all_targets(self):
 
         def _output_mlir(prog: ModuleOp, output: IO[str]):
             printer = Printer(
-                stream=output, print_generic_format=self.args.print_op_generic
+                stream=output,
+                print_generic_format=self.args.print_op_generic,
+                print_debuginfo=self.args.print_debuginfo
+                if "print_debuginfo" in self.args
+                else False,
```
Comment (math-fehr) [inline xdsl/xdsl_opt_main.py:336]: Can you add an option for `print_debuginfo` instead? So it appears in `--help` for instance?

**Review by math-fehr — APPROVED** [review]
Nice, that looks good!

**Comment by math-fehr** [comment]
Could you also duplicate your test to pass through MLIR? That way we make sure it works well with it!

File: tests/filecheck/dialects/builtin/parse_with_location.mlir
Comment (webmiche) [inline tests/filecheck/dialects/builtin/parse_with_location.mlir:1]: Could you use the testing dialect here instead? That way the builtin tests won't depend on the func and arith dialect.

File: tests/test_printer.py
Comment (webmiche) [inline tests/test_printer.py:1]: Same here.

File: tests/filecheck/dialects/builtin/parse_with_location.mlir [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// RUN: xdsl-opt %s --print-op-generic | filecheck %s
+// CHECK: module
+"builtin.module"() ({
+  "func.func"() ({
+    // CHECK: ^{{.*}}(%{{.*}}: i32):
```
Comment (superlopuh) [inline tests/filecheck/dialects/builtin/parse_with_location.mlir:5]: Am I understanding correctly that this checks only for parsing and not printing? Do we want a test for printing also?

**Review by superlopuh — CHANGES_REQUESTED** [review]
Would love full location support, thank you for starting this

File: xdsl/xdsl_opt_main.py (hunk 2 above)
Comment (Moxinilian) [reply xdsl/xdsl_opt_main.py:336]: I looked for this but have no idea where this is happening. Any pointer?

File: tests/filecheck/dialects/builtin/parse_with_location.mlir (hunk 3 above)
Comment (Moxinilian) [reply tests/filecheck/dialects/builtin/parse_with_location.mlir:5]: Tests for printing happen in `test_printer.py`. The reason they are done separately is that right now this printing and parsing is only here for compatibility purposes. It does not add any information to the IR.

File: tests/test_printer.py
Comment (Moxinilian) [reply tests/test_printer.py:1]: I'm not sure I understand. I copied the test above to make this test.

File: tests/filecheck/dialects/builtin/parse_with_location.mlir
Comment (Moxinilian) [reply tests/filecheck/dialects/builtin/parse_with_location.mlir:1]: What is the testing dialect?

Comment (superlopuh) [reply tests/filecheck/dialects/builtin/parse_with_location.mlir:1]: It's in the dialects folder in xdsl

Comment (webmiche) [reply tests/filecheck/dialects/builtin/parse_with_location.mlir:1]: The testing dialect is our way of making tests independent. Check for example [the affine tests](https://github.com/xdslproject/xdsl/blob/main/tests/filecheck/dialects/affine/affine_ops.mlir). Here we use the test dialect to create the SSAValues that are the operands of other ops and to get the types. Overall, the dialect is quite small ([here](https://github.com/xdslproject/xdsl/blob/main/xdsl/dialects/test.py)).

File: tests/test_printer.py
Comment (webmiche) [reply tests/test_printer.py:1]: This comment was about using the testing dialect. I know, this is not (yet) done in all the tests, but I think it makes sense to do it for new tests.

**Review by math-fehr — APPROVED** [review]
:+1:

**Comment by math-fehr** [comment]
@superlopuh, if that's good for you, we can merge it

File: tests/filecheck/dialects/builtin/parse_with_location.mlir [hunk 4]
```diff
@@ -0,0 +1,12 @@
+// RUN: xdsl-opt %s --print-op-generic | filecheck %s
+// CHECK: module
+"builtin.module"() ({
+  "func.func"() ({
```
Comment (webmiche) [inline tests/filecheck/dialects/builtin/parse_with_location.mlir:4]: Why not remove the functions as well?

**Review by AntonLydike — APPROVED** [review]

**Review by superlopuh — APPROVED** [review]
my b

**Review by webmiche — APPROVED** [review]
LGTM now 👍

**Pull request merged**